MEN-3240 propagate device status to inventory: use the wf

merlin-northern opened the pull request:
Propagate device status to inventory
--

* **deviceauth** calls _update_device_status_ [[1](https://github.com/mendersoftware/workflows/pull/40)] workflow to propagate information on device status to inventory.
* **inventory** exposes internal API to set status for a list of devices [[2](https://github.com/mendersoftware/inventory-enterprise/pull/24)].
* _/filters/search_ can be use by UI to get list of devices as a single place of information.

related PRs:
 * [1] https://github.com/mendersoftware/workflows/pull/40
 * [2] https://github.com/mendersoftware/inventory/pull/162
 * [3] https://github.com/mendersoftware/deviceauth/pull/324

ChangeLog:none
Signed-off-by: Peter Grzybowski <peter@northern.tech>

merlin-northern commented:
https://gitlab.com/Northern.tech/Mender/mender-qa/pipelines/136596505

merlin-northern commented on api/http/api_devauth.go:
```diff
@@ -52,6 +52,8 @@ const (
 	v2uriToken               = "/api/management/v2/devauth/tokens/:id"
 	v2uriDevicesLimit        = "/api/management/v2/devauth/limits/:name"
 
+	v3uriDeviceStatus = "/api/management/v3/devauth/devices/:status"
```
that is just an extra not-used-yet thing, I just added it since I believe this is the right way to proceed, and I was touching this anyway.

mchalski commented:
@merlin-northern I guess i need some fundamental clarifications:
- so the endpoint accepts a list of authsets - and it's public, how is this actually helping status propagation? my understanding was that on every status update devauth will post an update to inventory, why the public enpoint?
- related - I see you operate in bulk, on multiple IDs, can you clarify the scenario?

merlin-northern commented:
> @merlin-northern I guess i need some fundamental clarifications:
> 
> * so the endpoint accepts a list of authsets - and it's public, how is this actually helping status propagation? my understanding was that on every status update devauth will post an update to inventory, why the public enpoint?
> * related - I see you operate in bulk, on multiple IDs, can you clarify the scenario?

ok I can see that I have created a distraction and confusion with the extra (unused) endpoint I added. I have removed it, since it does not strictly belong to this PR.
all the update happens via workflow. I have also change the target branch to master.

mchalski commented on client/orchestrator/client_orchestrator.go:
```diff
@@ -58,6 +59,20 @@ type ProvisionDeviceReq struct {
+	// Request ID
+	RequestId string `json:"request_id"`
+	// Device IDs
+	//Ids []string `json:"device_ids"`
```
which is it then? what's up with the list of strings?

merlin-northern replied:
see below, I will remove the commented line.

mchalski commented on client/orchestrator/client_orchestrator.go:
```diff
@@ -58,6 +59,20 @@ type ProvisionDeviceReq struct {
+	RequestId string `json:"request_id"`
+	// Device IDs
+	//Ids []string `json:"device_ids"`
+	Ids string `json:"device_ids"`
```
it's either a single Id, or a list of Ids.

what's the backstory? in what scenario is it useful to update the status of multiple devices?

merlin-northern replied:
the back story is: for now we only update a single device. but we have to change the UI to update multiple devices at once. so the workflow supports that, but you have to pass to the workflow a valid list in json, since there is no place to Marshal the arguments in the workflows, this has to be Marshalled [here](https://github.com/mendersoftware/deviceauth/pull/324/commits/d55f727cfd92f5ed7328cab221616abc57c9d03c#diff-bafc5a98b88ba2baa09868cc001b23e0R473)
So the workflow is as follows:
1. device changes state
2. deviceauth creates one element array with the device id, marshals it into json string
3. deviceauth calls update_device_status workflow which passes the string from 2. containing json array of ids as a workflow [parameter](https://github.com/mendersoftware/inventory/pull/162/commits/1c00753a07a513366699ed866638060a788267aa#diff-d38bb1bb13c34ab0a761f45bb70cd491R623) into POST call to internal inventory [endpoint](https://github.com/mendersoftware/inventory/pull/162/commits/1c00753a07a513366699ed866638060a788267aa#diff-d38bb1bb13c34ab0a761f45bb70cd491R623).
At one point the device auth will have to be called from UI with multiple devices to update (that is what the endpoint I had added here was supposed to be doing.)
